src/server/scripts/Spells/spell_dh.cpp

```diff
@@ -507,35 +509,85 @@ class spell_dh_collective_anguish_eye_beam : public AuraScript
     }
 };
 
-// Called by 188499 - Blade Dance, 162794 - Chaos Strike, 185123 - Throw Glaive and 342817 - Glaive Tempest
+// Called by 198013 - Eye Beam
 class spell_dh_cycle_of_hatred : public SpellScript
 {
     bool Validate(SpellInfo const* /*spellInfo*/) override
     {
-        return ValidateSpellInfo({ SPELL_DH_CYCLE_OF_HATRED });
+        return ValidateSpellInfo({ SPELL_DH_CYCLE_OF_HATRED_TALENT, SPELL_DH_CYCLE_OF_HATRED_COOLDOWN_REDUCTION, SPELL_DH_CYCLE_OF_HATRED_REMOVE_STACKS });
     }
 
     bool Load() override
     {
-        if (!GetCaster()->HasAura(SPELL_DH_CYCLE_OF_HATRED))
-            return false;
+        return GetCaster()->HasAura(SPELL_DH_CYCLE_OF_HATRED_TALENT);
+    }
+
+    void HandleCycleOfHatred() const
+    {
+        Unit* caster = GetCaster();
+
+        // First calculate cooldown then add another stack
+        Aura const* cycleOfHatredStack = caster->GetAura(SPELL_DH_CYCLE_OF_HATRED_COOLDOWN_REDUCTION);
+        SpellInfo const* cycleOfHatred = sSpellMgr->AssertSpellInfo(SPELL_DH_CYCLE_OF_HATRED_TALENT, DIFFICULTY_NONE);
+        caster->GetSpellHistory()->ModifyCooldown(GetSpellInfo(), -Milliseconds(cycleOfHatred->GetEffect(EFFECT_0).CalcValue() * cycleOfHatredStack->GetStackAmount()));
 
-        if (GetSpellInfo()->Id != SPELL_DH_THROW_GLAIVE)
-            return true;
+        CastSpellExtraArgs args;
+        args.SetTriggerFlags(TRIGGERED_IGNORE_CAST_IN_PROGRESS | TRIGGERED_DONT_REPORT_CAST_ERROR);
+        args.SetTriggeringSpell(GetSpell());
 
-        // Throw Glaive triggers this talent only with Furious Throws
-        return GetCaster()->HasAura(SPELL_DH_FURIOUS_THROWS);
+        caster->CastSpell(caster, SPELL_DH_CYCLE_OF_HATRED_COOLDOWN_REDUCTION, args);
+        caster->CastSpell(caster, SPELL_DH_CYCLE_OF_HATRED_REMOVE_STACKS, args);
     }
 
-    void ReduceEyeBeamCooldown() const
+    void Register() override
+    {
+        AfterCast += SpellCastFn(spell_dh_cycle_of_hatred::HandleCycleOfHatred);
+    }
+};
+
+// 1214890 - Cycle of Hatred
+class spell_dh_cycle_of_hatred_remove_stacks : public AuraScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_DH_CYCLE_OF_HATRED_COOLDOWN_REDUCTION });
+    }
+
+    void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/) const
+    {
+        if (Aura* aura = GetTarget()->GetAura(SPELL_DH_CYCLE_OF_HATRED_COOLDOWN_REDUCTION))
```

Comment:
maybe using load to check the aura being applied and then just SetStackAmount to 1 OnRemove